snapshot_iterator.go

```diff
@@ -88,37 +89,69 @@ func (config *snapshotIteratorConfig) init() error {
 		return fmt.Errorf("tables is required")
 	}
 
+	if config.getMasterPos == nil {
+		return fmt.Errorf("getMasterPos mandatory")
+	}
+
 	return nil
 }
 
 func newSnapshotIterator(
 	ctx context.Context,
 	config snapshotIteratorConfig,
-) (common.Iterator, error) {
+) (*snapshotIterator, error) {
 	if err := config.init(); err != nil {
 		return nil, fmt.Errorf("invalid snapshot iterator config: %w", err)
 	}
 
-	// start position is mutable, so in order to avoid unexpected behaviour in
+	// Start position is mutable, so in order to avoid unexpected behaviour in
 	// tests we clone it.
 	lastPosition := config.startPosition.Clone()
 
 	iterator := &snapshotIterator{
-		t:            &tomb.Tomb{},
-		data:         make(chan fetchData),
-		acks:         csync.WaitGroup{},
-		config:       config,
+		t:    &tomb.Tomb{},
+		data: make(chan fetchData),
+		acks: csync.WaitGroup{}, config: config,
 		lastPosition: lastPosition,
 	}
 
+	// We first obtain all transaction locks for every table so that we can then
```

Comment:
If we are using debezium as an example, they use a global read lock see https://dev.mysql.com/doc/refman/8.4/en/flush.html#flush-tables-with-read-lock (there is an option to acquire a read lock for the specific tables as well). The way I see this play out:

1. Get global lock
2. Open tx; aquire the snapshot beginning and end values for each table
3. Get master position
4. Release lock

From what I can tell, only 2-3 is done, outside the safety of the read lock?

Reply:
Yeah sorry! I thought that with transactions only I had enough. Latest commit should add steps 1 and 4.